Core: Optimize manifest evaluation for super wide tables

irshadcc opened the pull request:
During the snapshot commit process of MergingSnapshotProducer, Iceberg tries to merge the manifest files to increase the planning performance. During operations like overwrite, MergingSnapshotProducer finds the matching manifest files by deleteExpression and rewrites the manifests by filtering out the deleted manifest entries.

https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L330

While finding the matching manifests by deleteExpression, we found that the Schema is created every time the expression needs to be bound. This has proven very expensive when there are large number of manifest files (~25,000 manifest files) for a super wide table (~35,000 columns).

https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/expressions/NamedReference.java#L41

For optimising the manifest evaluation, we can add a method called asSchema() in the StructType class to avoid creating the new Schema every time the filter needs to be bound.

irshadcc commented:
Resolves issue #9118

Fokko commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -592,6 +594,13 @@ public Types.StructType asStructType() {
       return this;
     }
 
+    public Schema asSchema() {
```
Could you add a Javadoc here to indicate that it is for optimization

Fokko reviewed (approved):
Thanks for raising this @irshadcc, this looks good to me. I've left two small comments, could you take a peek at those? Thanks for fixing this! 🙌

irshadcc commented:
> Thanks for raising this @irshadcc, this looks good to me. I've left two small comments, could you take a peek at those? Thanks for fixing this! 🙌

I've added the Javadoc and removed the empty.

irshadcc commented:
@Fokko Can we merge this PR ?

irshadcc commented:
kind ping @Fokko

Fokko commented:
Hey @irshadcc Sorry for the long wait here, and thanks for pinging me. Let's get this in 🚀
